include/oneapi/dpl/internal/scan_by_segment_impl.h

```diff
@@ -0,0 +1,372 @@
+                  _ValueType __accumulator, _BinaryOp __binary_op, ::std::size_t __wgroup_size)
+{
+    _IdxType __first = 0;
+    __local_acc[__local_id] = __accumulator;
+
+    __dpl_sycl::__group_barrier(__item);
+
+    for (::std::size_t __i = 1; __i < __wgroup_size; __i += __i)
+    {
+        if (__delta_local_id >= __i)
+            __accumulator = __binary_op(__local_acc[__first + __local_id - __i], __accumulator);
+
+        __first = __wgroup_size - __first;
+        __local_acc[__first + __local_id] = __accumulator;
+
+        __dpl_sycl::__group_barrier(__item);
+    }
+
+    return (__local_id ? __local_acc[__first + __local_id - 1] : 0);
+}
+
+enum class scan_type
+{
+    inclusive,
+    exclusive
+};
+
+template <scan_type __scan_type, typename... Name>
+class __sycl_segmented_scan_kernel1;
+
+template <scan_type __scan_type, typename... Name>
+class __sycl_segmented_scan_kernel2;
+
+template <scan_type __scan_type>
+struct sycl_scan_by_segment_impl
+{
+    template <typename... _Name>
+    using _KernelName1 = __sycl_segmented_scan_kernel1<__scan_type, _Name...>;
+
+    template <typename... _Name>
+    using _KernelName2 = __sycl_segmented_scan_kernel2<__scan_type, _Name...>;
+
+    template <typename _ExecutionPolicy, typename _Range1, typename _Range2, typename _Range3,
+              typename _BinaryPredicate, typename _BinaryOperator, typename _T>
+    void
+    sycl_scan_by_segment(_ExecutionPolicy&& __exec, _Range1&& __keys, _Range2&& __values, _Range3&& __out_values,
+                         _BinaryPredicate __binary_pred, _BinaryOperator __binary_op, _T __init, _T __identity)
+    {
+        using _Policy = typename ::std::decay<_ExecutionPolicy>::type;
+        using _CustomName = typename _Policy::kernel_name;
+
+        using _SegScanKernel1 = oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_generator<
+            _KernelName1, _CustomName, _Range1, _Range2, _Range3,  _BinaryPredicate, _BinaryOperator>;
+        using _SegScanKernel2 = oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_generator<
+            _KernelName2, _CustomName, _Range1, _Range2, _Range3,  _BinaryPredicate, _BinaryOperator>;
+
+        using __diff_type = oneapi::dpl::__internal::__difference_t<_Range1>;
+        using __key_type = oneapi::dpl::__internal::__value_t<_Range1>;
+        using __val_type = oneapi::dpl::__internal::__value_t<_Range2>;
+        using __flag_type = bool;
+
+        const ::std::size_t __n = __keys.size();
+
+        constexpr ::std::size_t __vals_per_item =
+            4; // Assigning 4 elements per work item resulted in best performance on gpu.
+
+        ::std::size_t __wgroup_size = oneapi::dpl::__internal::__max_work_group_size(__exec);
+
+        // change __wgroup_size according to local memory limit
+        __wgroup_size = oneapi::dpl::__internal::__max_local_allocation_size(::std::forward<_ExecutionPolicy>(__exec),
```

Comment:
`__max_local_allocation_size` should be replaced with `__slm_adjusted_work_group_size` since #811.

Reply:
I have updated this.